lib/ansible/modules/windows/win_optional_feature.ps1

```diff
@@ -0,0 +1,108 @@
+#!powershell
+
+# Copyright: (c) 2019, Carson Anderson <rcanderson23@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$spec = @{
+	options = @{
+		name = @{ type = "str"; required = $true }
+		state = @{ type = "str"; default = "present"; choices = @("absent", "present") }
+		source = @{ type = "str" }
+		include_parent = @{ type = "bool"; default = $false }
+	}
+	supports_check_mode = $true
+}
+
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
+$module.Result.rc = 0
+
+$name = $module.Params.name
+$state = $module.Params.state
+$source = $module.Params.source
+$include_parent = $module.Params.include_parent
+
+$win_version = (Get-ItemProperty "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\").ProductName
+if ($win_version -notlike "Windows 10*") {
+    $module.FailJson("This version of Windows does not support the Enable-WindowsOptionalFeature.")
+}
+
+$feature_state_start = Get-WindowsOptionalFeature -Online -FeatureName $name
+if (-not $feature_state_start) {
+    $module.Result.rc = 1
+    $module.FailJson("Failed to find feature.")
+}
+
+
+if ($state -eq "present") {
+    if ($feature_state_start.State -notlike "Enabled*") {
+        $install_args = @{
+            FeatureName = $name
+            All = $include_parent
+        }
+
+        if ($source) {
+            if (-not (Test-Path -Path $source)) {
+                $module.Result.rc = 1
+                $module.FailJson("Path could not be found")
+            }
+            $install_args.Source = $source
+        }
+
+        if ($module.CheckMode) {
```

Comment:
Prematurely existing when in check mode makes the code a bit more confusing, the better and more common way is to just skip the actual cmdlet invocation and exit at the bottom of the script.